src/compiler/program.ts

```diff
@@ -599,99 +636,106 @@ namespace ts {
             }
         }
 
-        function tryReuseStructureFromOldProgram(): boolean {
+        function tryReuseStructureFromOldProgram(): StructuralChangesFromOldProgram {
             if (!oldProgram) {
-                return false;
+                return StructuralChangesFromOldProgram.NoOldProgram;
             }
 
             // check properties that can affect structure of the program or module resolution strategy
             // if any of these properties has changed - structure cannot be reused
             const oldOptions = oldProgram.getCompilerOptions();
             if (changesAffectModuleResolution(oldOptions, options)) {
-                return false;
+                return StructuralChangesFromOldProgram.ModuleResolutionOptions;
             }
 
             Debug.assert(!oldProgram.structureIsReused);
 
             // there is an old program, check if we can reuse its structure
             const oldRootNames = oldProgram.getRootFileNames();
             if (!arrayIsEqualTo(oldRootNames, rootNames)) {
-                return false;
+                return StructuralChangesFromOldProgram.RootNames;
             }
 
             if (!arrayIsEqualTo(options.types, oldOptions.types)) {
-                return false;
+                return StructuralChangesFromOldProgram.TypesOptions;
             }
 
             // check if program source files has changed in the way that can affect structure of the program
             const newSourceFiles: SourceFile[] = [];
             const filePaths: Path[] = [];
             const modifiedSourceFiles: { oldFile: SourceFile, newFile: SourceFile }[] = [];
+            let structuralChanges = StructuralChangesFromOldProgram.None;
 
             for (const oldSourceFile of oldProgram.getSourceFiles()) {
                 let newSourceFile = host.getSourceFileByPath
                     ? host.getSourceFileByPath(oldSourceFile.fileName, oldSourceFile.path, options.target)
                     : host.getSourceFile(oldSourceFile.fileName, options.target);
 
                 if (!newSourceFile) {
-                    return false;
+                    structuralChanges |= StructuralChangesFromOldProgram.SourceFileRemoved;

```

Comment:
SourceFileRemoved is in your `CannotReuseResolution` set, so why collect flags and continue here rather than just return?